GAWB-486: optimize db query for loading workflows for a submission

davidangb opened the pull request:
Previously, the code made multiple db queries for each workflow in a submission. Now, we make a single query to populate all workflows in a submission.

This requires PO approval due to an API change: when GET-ting a submission, we no longer populate the input resolutions for each workflow in the submission. If you want the input resolutions, you need to GET the workflow object itself. This change does not affect the UI. (approved: Chet, David S)

I waffled a lot on the approach to removing input resolutions. Currently, the "inputResolutions" remains in the response, but it is always empty. This forced some awkward changes to tests. I also haven't changed swagger in any way, will do that once we agree on an approach.

The alternate approach is to remove the key entirely from the response and use a new case class e.g. "WorkflowList" instead of Workflow. I'll go this approach if necessary, but it involves significantly higher risk of regression.

This does NOT optimize workflow failures, and there is further work that could be done to reduce queries even more (e.g. better joins for the submission's config and entity). I settled here as a cost:benefit win; the further optimizations may - or may not! - be micro-optimizations.

For the use case detailed in the jira ticket, here's the count of db queries:
BEFORE: 18 overhead, 409 to populate submission
AFTER: 18 overhead, 9 to populate submission
( the 18 overhead queries handle permissions, primarily. Your count of overhead queries may vary based on your login. )
- [x] **Submitter**: Rebase to develop. DO NOT SQUASH
- [ ] **Submitter**: Make sure Swagger is updated if API changes
- [x] **Submitter**: Make sure documentation for code is complete
- [x] **Submitter**: Make sure liquibase is updated if appropriate
- [x] **Submitter**: Review code comments; remove done TODOs, create stories for remaining TODOs
- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Add description or comments on the PR explaining the hows/whys (if not obvious)
- [x] Tell ![](http://i.imgur.com/9dLzbPd.png) that the PR exists if he wants to look at it
- [x] Anoint a lead reviewer (LR). **Assign PR to LR**
- [x] **LR**: Initial review by LR and others.
- [ ] Comment / review / update cycle:
  - Rest of team may comments on PR at will
  - **LR assigns to submitter** for feedback fixes
  - Submitter updates documentation as needed
  - Submitter rebases to develop again if necessary
  - Submitter makes further commits. DO NOT SQUASH. **Reassign to LR** for further feedback
- [x] ![](http://i.imgur.com/9dLzbPd.png) sign off
- [x] **LR** sign off
- [x] **Assign to submitter** to finalize
- [x] **Submitter**: Squash commits, rebase if necessary
- [x] **Submitter**: Verify all tests go green, including CI tests
- [x] **Submitter**: Merge to develop 
- [x] **Submitter**: Delete branch after merge
- [x] **Submitter**: Check configuration files in Jenkins in case they need changes
- [x] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [x] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [x] **Submitter**: Inform other teams of any API changes via hipchat and/or email
- [x] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

davidangb committed:
optimize db query for loading workflows for a submission

abaumann commented on src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/SubmissionComponent.scala:
```diff
@@ -246,9 +245,27 @@ trait SubmissionComponent {
+
+      workflows.result.map{recs =>
+        recs.groupBy(_._1.id).map{x =>
+          val wr:WorkflowRecord = x._2.head._1
```
personally I get very confused by all the tuples (which may just be lack of slick experience), but comments would be great to know what some of this is doing.  Something like val wr:WorkflowRecord = x._2.head._1 is at least not too bad because the value shows what that means, but I don't quite know what x is or the x._2 is that gets mapped below to get y._3

abaumann commented on src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/SubmissionComponentSpec.scala:
```diff
@@ -75,7 +87,10 @@ class SubmissionComponentSpec extends TestDriverComponentWithFlatSpecAndMatchers
 
-    assertResult(Some(testData.submission1.copy(status = SubmissionStatuses.Done))) {
+    // for performance reasons, input resolutions are not returned when listing workflows inside a submission
+    val expectedWorkflows = testData.submission1.workflows.map{_.copy(inputResolutions=Seq.empty)}
```
Yeah agreed it's a bit awkward to have to do all this in the unit tests to support overloading the use of Workflow rather than making a different case class.  If we are fine on maintaining the tests this way or accruing the tech debt later after the fire, then this seems fine to me.

dvoet commented on src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/SubmissionComponent.scala:
```diff
@@ -246,9 +245,42 @@ trait SubmissionComponent {
+
+      workflows.result.map{recs =>
+        // group the workflow/entity/message records by workflowId, which creates a map keyed by workflowId
+        val recordsByWorkflowId: Map[Long, Seq[(WorkflowRecord, EntityRecord, Option[WorkflowMessageRecord])]] = recs.groupBy(_._1.id)
```
it is more readable to use a case statement instead of tupleese

 recs.groupBy { case (workflow, entity, message) => workflow.id }

then the comment above is not as necessary

dvoet commented on src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/SubmissionComponent.scala:
```diff
@@ -246,9 +245,42 @@ trait SubmissionComponent {
+        // group the workflow/entity/message records by workflowId, which creates a map keyed by workflowId
+        val recordsByWorkflowId: Map[Long, Seq[(WorkflowRecord, EntityRecord, Option[WorkflowMessageRecord])]] = recs.groupBy(_._1.id)
+        // now that we're grouped, process each workflow's records, translating into a Workflow object
+        recordsByWorkflowId.map{entry =>
```
again, instead of using a tuple, use a case
recordsByWorkflowId.map{ case (workflowId,  recordTuple) =>
will make the code below much more clear

dvoet commented:
I don't really like this api change. I realize that it matches the UI workflow but it feels like we are leaving information out of the API. How many more queries would it be to get this info? Or we could go the other way and not list anything other that what is on the workflow record.

davidangb commented:
Gotcha on cases vs. tuples; I'll do that. I'll also investigate the impact of putting input resolutions back into the response ... though I wish you had said that when I first proposed the change :)

dvoet commented:
my opinion changed upon seeing the code. I think we should either show all info nested under a workflow or none of it in this api call. What is proposed here is a mix which seems arbitrary and leads to a dirty api.

davidangb commented:
@abaumann @dvoet back to you ... I think I have addressed all the code style comments, and I have added input resolutions back into the API, with only 1-2 extra queries. So we should be down from 409 to 10 or 11.

abaumann commented on src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/SubmissionComponent.scala:
```diff
@@ -246,9 +245,73 @@ trait SubmissionComponent {
+      
+      inputResolutions.result.flatMap { resolutions =>
+        // first, prepare all the input resolutions
+        val resolutionsByWorkflowId = inputResolutions.result.map { resolutionRecs =>
```
you are doing a inputResolutions.result.map inside a inputResolutions.result.flatMap, it looks like this could occur once outside the flatMap?

davidangb replied:
Oooooh, good catch, thank you. Refactoring this cleaned up a few more lines of code, in the latest commit.

dvoet commented on src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/SubmissionComponent.scala:
```diff
@@ -246,9 +245,72 @@ trait SubmissionComponent {
     }
 
     def loadSubmissionWorkflows(submissionId: UUID): ReadAction[Seq[Workflow]] = {
```
this feels like it belongs in the WorkflowComponent but that can be for another day.

dvoet commented:
:+1:

abaumann commented:
👍
